### PR TITLE
Implementing a GraphQL API server using Go lang and a MongoDB database

### LINKED ISSUE #5927: [API]Implementing a GraphQL API server using Go lang and a MongoDB database
#### Proposed title of article

Implementing a GraphQL API server using Go lang and a MongoDB database

#### Proposed article introduction

GraphQL is a query language for APIs used to communicate data between a client and a server. This allows you to structure data-driven applications much more flexibly and efficiently than the REST AND SOAP approach. GraphQL provides a complete and understandable description of the data in your API. It gives clients the power to ask for exactly what they need and nothing more. This makes it easier to evolve APIs over time and enables powerful developer tools. It also gives developers more control over how they use data in their apps.

GraphQL can be implemented with different languages and databases of choice such as Golang and MongoDB. Go is a general-purpose language meaning that it can be used for many different things. You can use Go to build web apps, microservices, cloud services, APIs, DevOps toolings, and any application you can think of. This means you will probably use Go to build a server-side-based application. Just like Node.js, you will need the right tooling to be able to handle your server-side code.

After developing your application with Go, you might need a database to store data. One of the databases you can use is MongoDB which is a cross-platform NoSQL database. MongoDB is a cross-platform document-oriented database platform, classified as a NoSQL database. It uses JSON objects as its data tuples. When working with a web application, you might not be aware of the exact data format being sent, and a NoSQL database such a MongoDB would be a good solution for data handling and storage.

This guide runs a GraphQL server using Golang and a MongoDB database.

#### Key takeaways

Introduction to GraphQL
What is GraphQL
Why GraphQL APIs
Configuring a GraphQL server in Go
How to use Mutations and Queries in GraphQL
Create a Golang GraphQL server-based application with all CRUD operations supported

#### Article quality

A number of articles explain the concept of GraphQL with the likes of MERN and MEVN stack. However, GraphQL uses a mechanism that queries data and then displays that data with the GraphQL schema. This guide aims to introduce the Go programmers to the concept of GraphQL. We will explain this concept and how it works. The build and application stack that has GraphQL, Golang, and MongoDB. The reader will also understand the frameworks and libraries we will use to achieve this.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Joakim-gakure**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [GitHub Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5927

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by srishilesh** [comment]
Hello @Joakim-gakure ,

Your PR has entered our reviewing queue. Please wait until a peer-reviewer pings you to get started with the review process.

To avoid potential closure of the PR:
- Perform a grammar check
- Make sure the hero image is of dimension 600x400
- Start markdown headers from `###`
- Go through the markdown guidelines that we follow here, by referring to our contribution guidelines
- For using bulleted points, use `-` instead of `*`
- For code snippets, make sure to explain them properly or add comments wherever needed
- The image paths are correct
- The author info is correct (Refer to the guidelines to view the latest update in author info section)
- The directory names are in lower case, separated by `-`

We also suggest you to go through other approved PRs to see how the authors have formatted their articles.

Thanks.

**Comment by ahmadmardeni1** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. @Joakim-gakure 
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Implementing a GraphQL API server using Go lang and a MongoDB database#6077](https://github.com/section-engineering-education/engineering-education/files/7890711/New.Scan.4_19.PM.Copyleaks.Report.pdf)

**Comment by ahmadmardeni1** [comment]
Please unassign this PR @Jethro-magaji - You can only review 2/3 PRs at the same time.

**Comment by Jethro-magaji** [comment]
> Please unassign this PR @Jethro-magaji - You can only review 2/3 PRs at the same time.

Alright, done ✅.

**Comment by srishilesh** [comment]
> Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. @Joakim-gakure Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.
> 
> [Implementing a GraphQL API server using Go lang and a MongoDB database#6077](https://github.com/section-engineering-education/engineering-education/files/7890711/New.Scan.4_19.PM.Copyleaks.Report.pdf)

@Joakim-gakure 

Are the plagiarism changes done?

If you still need more time, you can work on your forked repository and create a new PR when done.

**Comment by Joakim-gakure** [comment]
Yes @srishilesh, Am adding the changes ASAP.

**Comment by Joakim-gakure** [comment]
Changes done @srishilesh and @ahmadmardeni1

**Comment by srishilesh** [comment]
@Joakim-gakure 

Please wait until a peer-reviewer pings you with the changes to be made.

@section-engineering-education/peer-reviewers Is anyone available to review this PR?

**Comment by ahmadmardeni1** [comment]
Hey @Bonimum  We have filled our peer reviewers team. You will be considered when we have open positions.

File: content/articles/implement-graphql-api-using-golang-and-mongodb-database/index.md [hunk 1]
```diff
@@ -327,15 +332,15 @@ import (
 ```
 
 ### Set the Go server
-To run this GraphQL API, we need to expose the methods we have created to localhost so that we can access them using an endpoint. To implement this, follow the following steps.
+To run this GraphQL API, we need to expose the methods we have created to localhost to access them using an endpoint. To implement this, follow the following steps.
```
Comment (Jethro-magaji) [inline content/articles/implement-graphql-api-using-golang-and-mongodb-database/index.md:335]: There's a typo here "we have created to localhost to ".

File: content/articles/implement-graphql-api-using-golang-and-mongodb-database/index.md [hunk 2]
```diff
@@ -423,7 +428,7 @@ query Movie{
 
 Where `:id` is the `_id` value of the movie you want your GraphQL API to return.
 
-- Getting all movies
+- Getting all movies:
 
 ```go
 query Movies{
```
Comment (Jethro-magaji) [inline content/articles/implement-graphql-api-using-golang-and-mongodb-database/index.md:434]: Your conclusion section should include any next or final steps, summarizing the tasks the audience just completed with possible next steps.

**Review by Jethro-magaji — CHANGES_REQUESTED** [review]
Hi! @Joakim-gakure 
Well done for such an amazing article, please go over my reviews and make the changes, I recommend editing your article with Grammarly or ProWriterAid just to double-check for typos or errors. 
Use "You" or "Your" instead of "We" or "Lets" in some instances.
Thanks.

**Comment by hectorkambow** [comment]
@Joakim-gakure 
Good afternoon and thank you for submitting your topic suggestion. Not a problem but it seems that this PR is still quite a bit aways from being completed. 
 
In an effort to be fair to our larger EngEd community as a rule of thumb - we can only accept content (pull requests) that are complete and fully polished articles/tutorial - meaning they are ready to be published. 
 
All revisions and edits **should be completed on your own forked repo** (as to not take up room in the queue) - this allows all revisions to be done on your forked personal repo.

**Review by Jethro-magaji — APPROVED** [review]
Hi! @Joakim-gakure 
Well done for coming up with such an awesome article. Please take note of the reviews made here and consider them next time you submit an article to minimize revision. 
Thanks.

**Comment by WanjaMIKE** [comment]
Nice work guys @Joakim-gakure  cc @Jethro-magaji

**Comment by WanjaMIKE** [comment]
Merged #6077 into master

**Pull request merged**